File: src/main/java/zwylair/randomitembattle/commands/StartposCommands.java
package zwylair.randomitembattle.commands;

import com.mojang.brigadier.CommandDispatcher;
import com.mojang.brigadier.context.CommandContext;

import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.server.command.ServerCommandSource;
import static net.minecraft.server.command.CommandManager.literal;
import net.minecraft.text.Text;
import net.minecraft.util.math.Vec3d;

import static zwylair.randomitembattle.RandomItemBattle.*;
import static zwylair.randomitembattle.utils.PosCalculator.roundPosition;


public class StartposCommands {

    public static void register(CommandDispatcher<ServerCommandSource> dispatcher) {
        dispatcher.register(literal("rib").then(literal("startpos").then(literal("add")
                .executes(StartposCommands::addStartpos))));

        dispatcher.register(literal("rib").then(literal("startpos").then(literal("remove")
                .executes(StartposCommands::removeStartpos))));

        dispatcher.register(literal("rib").then(literal("startpos").then(literal("clear")
                .executes(StartposCommands::removeAllStartpos))));

        dispatcher.register(literal("rib").then(literal("set_center_position")
                .executes(StartposCommands::setCenterPos)));
    }

    public static int addStartpos(CommandContext<ServerCommandSource> ctx) {
        PlayerEntity player = ctx.getSource().getPlayer();
        assert player != null;  // executor was not a player
        Vec3d spawnPosition = roundPosition(player.getPos());

        if (playerPositions.contains(spawnPosition)) {
            ctx.getSource().sendFeedback(() -> Text.literal(chatModPrefix + "§cThis position already added"), false);
            return 1;
        }
        playerPositions.add(spawnPosition);

        int x = (int) spawnPosition.x;
        int z = (int) spawnPosition.z;
        if (spawnPosition.x < 0) { x -= 1; }
        if (spawnPosition.z < 0) { z -= 1; }
        int finalX = x;
        int finalZ = z;

        ctx.getSource().sendFeedback(() -> Text.literal(chatModPrefix + "Position [x:%d, y:%d, z:%d] was added".formatted(finalX, (int) spawnPosition.y, finalZ)), false);
        return 0;
    }

    public static int removeStartpos(CommandContext<ServerCommandSource> ctx) {
        PlayerEntity player = ctx.getSource().getPlayer();
        assert player != null;  // executor was not a player
        Vec3d spawnPosition = roundPosition(player.getPos());

        if (!playerPositions.contains(spawnPosition)) {
            ctx.getSource().sendFeedback(() -> Text.literal(chatModPrefix + "§cPosition [x:%d, y:%d, z:%d] is not in position list".formatted((int) spawnPosition.x, (int) spawnPosition.y, (int) spawnPosition.z)), false);
            return 1;
        }
        playerPositions.remove(spawnPosition);

        int x = (int) spawnPosition.x;
        int z = (int) spawnPosition.z;
        if (spawnPosition.x < 0) { x -= 1; }
        if (spawnPosition.z < 0) { z -= 1; }
        int finalX = x;
        int finalZ = z;

        ctx.getSource().sendFeedback(() -> Text.literal(chatModPrefix + "Position [x:%d, y:%d, z:%d] was removed".formatted(finalX, (int) spawnPosition.y, finalZ)), false);
        return 0;
    }

    public static int removeAllStartpos(CommandContext<ServerCommandSource> ctx) {
        playerPositions.clear();

        ctx.getSource().sendFeedback(() -> Text.literal(chatModPrefix + "Start positions were removed"), false);
        return 0;
    }

    public static int setCenterPos(CommandContext<ServerCommandSource> ctx) {
        PlayerEntity player = ctx.getSource().getPlayer();
        assert player != null;  // executor was not a player
        centerPosition = roundPosition(player.getPos());

        ctx.getSource().sendFeedback(() -> Text.literal(chatModPrefix + "Center position was set"), false);
        return 0;
    }
}

File: src/main/kotlin/zwylair/randomitembattle/commands/StartposCommands.kt
package zwylair.randomitembattle.commands

import com.mojang.brigadier.CommandDispatcher
import com.mojang.brigadier.context.CommandContext
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.server.command.CommandManager.literal
import net.minecraft.server.command.ServerCommandSource
import net.minecraft.text.Text
import net.minecraft.util.math.Vec3d
import zwylair.randomitembattle.RandomItemBattle
import zwylair.randomitembattle.utils.PosCalculator.Companion.roundPosition

object StartposCommands {
    fun register(dispatcher: CommandDispatcher<ServerCommandSource>) {
        dispatcher.register(literal("rib").then(literal("startpos").then(literal("add").executes { addStartpos(it) })))
        dispatcher.register(literal("rib").then(literal("startpos").then(literal("remove").executes { removeStartpos(it) })))
        dispatcher.register(literal("rib").then(literal("startpos").then(literal("clear").executes { removeAllStartpos(it) })))
        dispatcher.register(literal("rib").then(literal("set_center_position").executes { setCenterPos(it) }))
    }

    private fun addStartpos(ctx: CommandContext<ServerCommandSource>): Int {
        val player: PlayerEntity = ctx.source.player ?: return 1  // return if executor was not a player
        val spawnPosition: Vec3d = roundPosition(player.pos)

        if (spawnPosition in RandomItemBattle.playerPositions) {
            ctx.source.sendFeedback({ Text.of("${RandomItemBattle.CHAT_PREFIX_MOD} §cThis position already added") }, false)
            return 1
        }

        RandomItemBattle.playerPositions.add(spawnPosition)

        val finalX = if (spawnPosition.x < 0) spawnPosition.x.toInt() - 1 else spawnPosition.x.toInt()
        val finalZ = if (spawnPosition.z < 0) spawnPosition.z.toInt() - 1 else spawnPosition.z.toInt()

        ctx.source.sendFeedback({ Text.of("${RandomItemBattle.CHAT_PREFIX_MOD} Position [x:$finalX, y:${spawnPosition.y.toInt()}, z:$finalZ] was added") }, false)
        return 0
    }

    private fun removeStartpos(ctx: CommandContext<ServerCommandSource>): Int {
        val player: PlayerEntity = ctx.source.player ?: return 1
        val spawnPosition: Vec3d = roundPosition(player.pos)
        val finalX = if (spawnPosition.x < 0) spawnPosition.x.toInt() - 1 else spawnPosition.x.toInt()
        val finalZ = if (spawnPosition.z < 0) spawnPosition.z.toInt() - 1 else spawnPosition.z.toInt()

        if (spawnPosition !in RandomItemBattle.playerPositions) {
            ctx.source.sendFeedback({ Text.of("${RandomItemBattle.CHAT_PREFIX_MOD} §cPosition [x:$finalX, y:${spawnPosition.y.toInt()}, z:$finalZ] is not in position list") }, false)
            return 1
        }
        RandomItemBattle.playerPositions.remove(spawnPosition)

        ctx.source.sendFeedback({ Text.of("${RandomItemBattle.CHAT_PREFIX_MOD} Position [x:$finalX, y:${spawnPosition.y.toInt()}, z:$finalZ] was removed") }, false)
        return 0
    }

    private fun removeAllStartpos(ctx: CommandContext<ServerCommandSource>): Int {
        RandomItemBattle.playerPositions.clear()
        ctx.source.sendFeedback({ Text.of("${RandomItemBattle.CHAT_PREFIX_MOD} Start positions were removed") }, false)
        return 0
    }

    private fun setCenterPos(ctx: CommandContext<ServerCommandSource>): Int {
        val player: PlayerEntity = ctx.source.player ?: return 1  // return if executor was not a player

        RandomItemBattle.centerPosition = roundPosition(player.pos)
        ctx.source.sendFeedback({ Text.of("${RandomItemBattle.CHAT_PREFIX_MOD} Center position was set") }, false)
        return 0
    }
}
